File: libs/document-processor-next/src/types/document-processing-result.ts
import type { BlueNode } from '@blue-labs/language';

export interface DocumentProcessingResult {
  readonly document: BlueNode;
  readonly triggeredEvents: readonly BlueNode[];
  readonly totalGas: number;
  readonly capabilityFailure: boolean;
  readonly failureReason: string | null;
}

function freezeEvents(events: readonly BlueNode[]): readonly BlueNode[] {
  return Object.freeze([...events]);
}

const factories = {
  of(
    document: BlueNode,
    triggeredEvents: readonly BlueNode[],
    totalGas: number
  ): DocumentProcessingResult {
    return {
      document,
      triggeredEvents: freezeEvents(triggeredEvents),
      totalGas,
      capabilityFailure: false,
      failureReason: null,
    };
  },

  capabilityFailure(
    document: BlueNode,
    reason: string | null
  ): DocumentProcessingResult {
    return {
      document,
      triggeredEvents: Object.freeze([]),
      totalGas: 0,
      capabilityFailure: true,
      failureReason: reason ?? null,
    };
  },
} as const;

export const DocumentProcessingResult = factories;

File: references/java-sdk/src/main/java/blue/language/processor/DocumentProcessingResult.java
package blue.language.processor;

import blue.language.model.Node;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

/**
 * Immutable value object representing the outcome of a single PROCESS run.
 */
public final class DocumentProcessingResult {

    private final Node document;
    private final List<Node> triggeredEvents;
    private final long totalGas;
    private final boolean capabilityFailure;
    private final String failureReason;

    private DocumentProcessingResult(Node document,
            List<Node> triggeredEvents,
            long totalGas,
            boolean capabilityFailure,
            String failureReason) {
        this.document = document;
        this.triggeredEvents = Collections.unmodifiableList(new ArrayList<>(triggeredEvents));
        this.totalGas = totalGas;
        this.capabilityFailure = capabilityFailure;
        this.failureReason = failureReason;
    }

    public static DocumentProcessingResult of(Node document, List<Node> triggeredEvents, long totalGas) {
        Objects.requireNonNull(document, "document");
        Objects.requireNonNull(triggeredEvents, "triggeredEvents");
        return new DocumentProcessingResult(document, new ArrayList<>(triggeredEvents), totalGas, false, null);
    }

    public static DocumentProcessingResult capabilityFailure(Node document, String reason) {
        Objects.requireNonNull(document, "document");
        return new DocumentProcessingResult(document, Collections.emptyList(), 0L, true, reason);
    }

    public Node document() {
        return document;
    }

    public List<Node> triggeredEvents() {
        return triggeredEvents;
    }

    public long totalGas() {
        return totalGas;
    }

    public boolean capabilityFailure() {
        return capabilityFailure;
    }

    public String failureReason() {
        return failureReason;
    }
}